### PR TITLE
core(fr): collect OOPIF network data

### PR DESCRIPTION
**Pull request by patrickhulce**
**Summary**
Adds OOPIF support to Fraggle Rock. ~~I'm still working on tests, but I lost too many days to brittle refactors, so I'm just putting up this as-is for feedback.~~ Bringing support here involved a number of components...

- Bump puppeteer dependency (and devtools-protocol to match) to get [`connection()` support](https://github.com/puppeteer/puppeteer/pull/6863).
- Expose `sessionattached` events on `FRProtocolSession`.
- Emit `sessionId` from `targetInfo` data on each session protocol event.
- Create a `TargetManager` class for managing the many sessions/targets we have going on.
- Update `NetworkMonitor` to leverage `TargetManager` for its network listeners.
- Update `DevtoolsLog` gatherer to use `NetworkMonitor` to avoid duplication of this complex logic.
- Update oopif smoketest to force the inclusion of iframe-elements audit (in Fraggle Rock unused artifacts are filtered out of the config)

**Related Issues/PRs**
ref #12861 #11313

### CODE REVIEW

**Comment by patrickhulce** [comment]
friendly ping here :)

File: lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js [hunk 1]
```diff
@@ -30,6 +30,32 @@ global.require = originalRequire;
...
+      // Just a string, is adding a new audit that won't be in the bundle, prune it.
+      if (typeof audit === 'string') return false;
+      // If it's the audit implementation directly, it's a new audit, prune it. (though not currently possible)
+      if (typeof audit !== 'object') return false;
+      // Keep the audit only if it is setting options, which is the signal that it's an existing core audit.
+      return Boolean(audit.options);
+    });
+  }
+
+  if (configJson.categories) {
```
Comment (connorjclark) [inline lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js:53]: what's going on here? does this not break `form-config` / form smoke?

File: lighthouse-cli/test/smokehouse/test-definitions/oopif/oopif-config.js [hunk 2]
```diff
@@ -11,6 +11,16 @@
...
   extends: 'lighthouse:default',
+  categories: {
+    performance: {
+      title: 'Performance',
+      auditRefs: [{id: 'iframe-elements', weight: 0}],
+    },
+  },
+  audits: [
+    // Include an audit that *forces* the IFrameElements artifact to be used for our test.
+    {path: `${__dirname}/oopif-iframe-audit.js`},
```
Comment (connorjclark) [inline lighthouse-cli/test/smokehouse/test-definitions/oopif/oopif-config.js:22]: LH_ROOT plz

File: lighthouse-core/gather/driver/target-manager.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        autoAttach: true,
+        flatten: true,
+        waitForDebuggerOnStart: true,
+      });
+    };
+  }
+
+  /**
+    * @param {LH.Gatherer.FRProtocolSession} session
+    */
```
Comment (connorjclark) [inline lighthouse-core/gather/driver/target-manager.js:48]: spacing

File: lighthouse-core/test/fraggle-rock/gather/mock-driver.js [hunk 4]
```diff
@@ -128,6 +159,20 @@ function mockDriverModule(driverProvider) {
   };
 }
 
+function mockTargetManagerModule() {
+  const targetManagerMock = createMockTargetManager();
+
+  /** @type {(instance: any) => (...args: any[]) => any} */
+  const proxyCtor = instance => function() {
+    // IMPORTANT! This must be a `function` not an arrow function so it can be invoked as a constructor.
```
Comment (connorjclark) [inline lighthouse-core/test/fraggle-rock/gather/mock-driver.js:167]: is this comment necessary or useful to future readers?

File: lighthouse-core/test/gather/driver/target-manager-test.js [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        .mockResponse('Target.getTargetInfo', {targetInfo})
+        .mockResponse('Target.setAutoAttach');
+      await targetManager.enable();
+
+      const invocations = sendCommandMock.findAllInvocations('Target.setAutoAttach');
+      expect(invocations).toHaveLength(0);
+    });
+
+    it('should fire listeners before target attached', async () => {
+
```
Comment (connorjclark) [inline lighthouse-core/test/gather/driver/target-manager-test.js:103]: empty :)

File: lighthouse-core/test/gather/mock-commands.js [hunk 6]
```diff
@@ -63,7 +63,7 @@ function createMockSendCommandFn(options) {
 
       const indexOfResponse = mockResponses
         .findIndex(entry => entry.command === command && entry.sessionId === sessionId);
-      if (indexOfResponse === -1) throw new Error(`${command} unimplemented`);
+      if (indexOfResponse === -1) return Promise.reject(new Error(`${command} unimplemented`));
```
Comment (connorjclark) [inline lighthouse-core/test/gather/mock-commands.js:66]: should we just make this function async?

**Review by connorjclark — APPROVED** [review]

File: lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js (hunk 1 above)
Comment (patrickhulce) [reply lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js:53]: form-config is disabled because it doesn't work without the flag :) 

Other options I see:

- hardcode this function to exclude only this oopif audit
- Add this as a default audit
- Include this audit in the DT bundle
- Test a different bundle than the DT bundle that does include this audit

Any preferences?

File: lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js [hunk 7]
```diff
@@ -30,6 +30,32 @@ global.require = originalRequire;
 // @ts-expect-error - not worth giving test global an actual type.
 const lighthouse = global.runBundledLighthouse;
 
+/**
```
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js:33]: this seems like a lot of workaround for

> in Fraggle Rock unused artifacts are filtered out of the config

if the goal is for legacy and FR to match soon, it seems like this is the thing to resolve rather than adding a bunch of new test stuff to support both? Downside is I found the two halves to make it work really hard to follow :)

If we reallllllly want to put off that decision, is this issue resolvable by injecting `oopif-iframe-audit` into the dt bundle?

File: lighthouse-core/test/fraggle-rock/gather/mock-driver.js (hunk 4 above)
Comment (patrickhulce) [reply lighthouse-core/test/fraggle-rock/gather/mock-driver.js:167]: IMO, yes. I can easily see a future mass change coming in and flipping this back without this comment

File: lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js (hunk 1 above)
Comment (brendankenny) [reply lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js:53]: ha, sorry, missed that you had already replied

File: lighthouse-core/test/gather/driver/target-manager-test.js (hunk 5 above)
Comment (patrickhulce) [reply lighthouse-core/test/gather/driver/target-manager-test.js:103]: haha thanks :)

File: lighthouse-core/test/gather/mock-commands.js (hunk 6 above)
Comment (patrickhulce) [reply lighthouse-core/test/gather/mock-commands.js:66]: done 👍

File: lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js (hunk 7 above)
Comment (patrickhulce) [reply lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js:33]: > it seems like this is the thing to resolve rather than adding a bunch of new test stuff to support both?

I'm not sure how this solves our problem given that we wanted to support unused artifact filtering for awhile. Aligning on this removes the `_skipInBundled` logic but still has this underlying issue that iframeelements is unused.

> If we reallllllly want to put off that decision, is this issue resolvable by injecting oopif-iframe-audit into the dt bundle?

Yes that's one of the options I outlined, is that your favorite one?

File: lighthouse-core/gather/driver/target-manager.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  async _onSessionAttached(session) {
+    try {
+      const target = await session.sendCommand('Target.getTargetInfo').catch(() => null);
+      const targetType = target && target.targetInfo && target.targetInfo.type;
+      const hasValidTargetType = targetType === 'page' || targetType === 'iframe';
+      if (!target || !hasValidTargetType) return;
+
+      const targetId = target.targetInfo.targetId;
+      session.setTargetInfo(target.targetInfo);
```
Comment (brendankenny) [inline lighthouse-core/gather/driver/target-manager.js:57]: could `session` take care of this itself rather than requiring an external class to query and set the targetInfo?

File: lighthouse-core/gather/driver/target-manager.js [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+      if (this._targets.has(targetId)) return;
+
+      const targetName = target.targetInfo.url || target.targetInfo.targetId;
+      log.verbose('target-manager', `target ${targetName} attached`);
+
+      const targetWithSession = {target: target.targetInfo, session};
+      this._targets.set(targetId, targetWithSession);
```
Comment (brendankenny) [inline lighthouse-core/gather/driver/target-manager.js:64]: is `targetWithSession` in this map eventually needed or could it just be a `Set` of targetIds?

Comment (brendankenny) [reply lighthouse-core/gather/driver/target-manager.js:64]: and is the plan still to eventually have multiple sessions per target?

File: lighthouse-core/gather/driver/target-manager.js [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+const log = require('lighthouse-logger');
+
+/** @typedef {{target: LH.Crdp.Target.TargetInfo, session: LH.Gatherer.FRProtocolSession}} TargetWithSession */
+
+class TargetManager {
+  /** @param {LH.Gatherer.FRProtocolSession} session */
+  constructor(session) {
+    this._enabled = false;
+    this._session = session;
```
Comment (brendankenny) [inline lighthouse-core/gather/driver/target-manager.js:21]: it feels like if the intention is we have a root session that maybe we should start naming it something special since it's not like the other sessions, but no rush to even think about that right now.

File: lighthouse-core/gather/driver/network-monitor.js [hunk 11]
```diff
@@ -49,14 +61,29 @@ class NetworkMonitor extends EventEmitter {
     this.off = (event, listener) => super.off(event, listener);
   }
 
+  /**
+   * @param {{target: {targetId: string}, session: LH.Gatherer.FRProtocolSession}} session
+   */
+  async _onTargetAttached({session, target}) {
+    const targetId = target.targetId;
+    if (this._sessions.has(targetId)) return;
```
Comment (brendankenny) [inline lighthouse-core/gather/driver/network-monitor.js:69]: won't the targetManager already be ensuring this won't be called twice with the same targetId?

**Review by brendankenny — COMMENTED** [review]
the kind of Lighthouse PR we haven't seen around here for some time :)

Added some exploratory comments/questions to see if I understand the flow. May be worth some ascii diagrams at some point :)

File: lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js (hunk 7 above)
Comment (brendankenny) [reply lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js:33]: > . Aligning on this removes the _skipInBundled logic but still has this underlying issue that iframeelements is unused.

I guess implicit in that is that any PR to make legacy filtered by default would have to fix the problem as well :)

> Yes that's one of the options I outlined, is that your favorite one?

if we were already filtering by default when we landed the iframeelements artifact, it does feel like we'd either have done that or added an outright pubads smoke test (since that's also already in the dt bundle). I'm cool with adding the test audit to the bundle. It could even be the clearinghouse audit specifically for including artifacts not being used by the default audits :)

File: lighthouse-core/gather/driver/network-monitor.js [hunk 12]
```diff
@@ -31,11 +35,19 @@ class NetworkMonitor extends EventEmitter {
     super();
     this._session = session;
 
+    this._messageLog = new MessageLog(/^(Page|Network)\./);
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/gather/driver/network-monitor.js:38]: the page events are largely network related (lifecycle networkidle) and aren't even used, so I wasn't very concerned about the extra bit. if folks feel it's a candidate for confusion, I can `emit` the messages and have the gatherer merge with page events instead, but that extra level of indirection felt unnecessary

File: lighthouse-core/gather/driver/target-manager.js (hunk 8 above)
Comment (patrickhulce) [reply lighthouse-core/gather/driver/target-manager.js:57]: it could, although there is no async initialization lifecycle on sessions today. having an async `init` that could fail to be called felt similarly brittle while also violating the lightweight connection invariant we have currently. passing the targetInfo in at the constructor was the closest I'm certainly up for, but that's still an external class querying and setting the targetInfo if that's where your concern is.

File: lighthouse-core/gather/driver/target-manager.js (hunk 9 above)
Comment (patrickhulce) [reply lighthouse-core/gather/driver/target-manager.js:64]: > and is the plan still to eventually have multiple sessions per target?

yep

> is targetWithSession in this map eventually needed or could it just be a Set of targetIds?

I toyed around with versions that exposed the targets to consumers which is why they're here, and I think we'll eventually do that for creating new sessions/exploring targets, but if it's concerning we can switch to just the set.

File: lighthouse-core/gather/driver/network-monitor.js (hunk 11 above)
Comment (patrickhulce) [reply lighthouse-core/gather/driver/network-monitor.js:69]: yes in this version I don't believe this line is necessary anymore 👍

File: lighthouse-core/gather/driver/network-monitor.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/gather/driver/network-monitor.js:38]: The restructuring seems totally fine, I was just talking about naming.

> the page events are largely network related (lifecycle networkidle) and aren't even used

but now `DevtoolsLog` the Gatherer gets _the_ devtools log from `NetworkMonitor`. A rename might be in order?

[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/gather/driver/network-monitor.js:38]: gotcha 👍 protocol-monitor SGTM

File: lighthouse-core/gather/driver/target-manager.js (hunk 8 above)
Comment (brendankenny) [reply lighthouse-core/gather/driver/target-manager.js:57]: > it could, although there is no async initialization lifecycle on sessions today

my point was that if this is the only use of subsession initialization and it's async to any observer of target-manager, might as well encapsulate `session` state in `session`, but sounds like that might be premature and the shape of things is still developing.

File: lighthouse-core/gather/driver/target-manager.js (hunk 9 above)
Comment (brendankenny) [reply lighthouse-core/gather/driver/target-manager.js:64]: > I toyed around with versions that exposed the targets to consumers which is why they're here, and I think we'll eventually do that for creating new sessions/exploring targets, but if it's concerning we can switch to just the set.

just live blogging me reading through the code and reasoning through what things are used for what. `Set` is probably the right thing to use for the current code, but it's not like there's going to be some deep YAGNI penalty for using a map with unused (for now) values over a set :)

**Review by brendankenny — APPROVED** [review]
not asked for, but just to be sure I'm not blocking anything: LGTM :) I'm looking forward to seeing how session handling continues to evolve

File: lighthouse-core/audits/oopif-iframe-audit.js [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * This mismatch in environment demands that the audit path in the config must be context-aware,
+ * yet the require-graph for the config is included before even the CLI knows that it will be using
+ * a bundled runner. Rather than force a massive smoketest architecture change, we include a harmless,
+ * test-only audit in our core list instead.
+ */
+
+module.exports = {
+  meta: {
+    id: 'oopif-iframe-audit',
```
Comment (connorjclark) [inline lighthouse-core/audits/oopif-iframe-audit.js:23]: `oopif-iframe-test-audit` ?

File: lighthouse-core/gather/driver/network-monitor.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/gather/driver/network-monitor.js]: > gotcha 👍 protocol-monitor SGTM

lol I take it you changed your mind? :)

[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/gather/driver/network-monitor.js]: yeah I got in there and it just didn't feel right, everything is so network-focused.

just reemitting the messages and letting the log continue to be captured by the gatherer :) (also bonus we get strict event listener types for the monitor now)

File: lighthouse-core/audits/oopif-iframe-audit.js (hunk 13 above)
Comment (patrickhulce) [reply lighthouse-core/audits/oopif-iframe-audit.js:23]: sure sg
